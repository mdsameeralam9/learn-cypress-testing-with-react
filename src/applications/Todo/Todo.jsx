import { useState } from "react";
//import {sortTodo} from "./helper/sort"

function TodoApp() {
  const [todos, setTodos] = useState([]);
  const [input, setInput] = useState("");
  const [editId, setEditId] = useState(null);
  const [editText, setEditText] = useState("");

  const addTodo = () => {
    if (input.trim()) {
      //setTodos(sortTodo([...todos, { id: Date.now(), text: input, completed: false }]))
      if (!editId) {
        setTodos([...todos, { id: Date.now(), text: input, completed: false }]);
      } else {
        setTodos((p) =>
          p.map((todo) =>
            todo.id === editId ? { ...todo, text: input } : todo
          )
        );
      }

      setInput("");
    }
  };

  const deleteTodo = (id) => {
    setTodos(todos.filter((todo) => todo.id !== id));
  };

  const toggleComplete = (id) => {
    setTodos(
      todos.map((todo) =>
        todo.id === id ? { ...todo, completed: !todo.completed } : todo
      )
    );
  };

  const startEdit = (id, text) => {
    setEditId(id);
    setEditText(text);
    setInput(text);
  };

  const saveEdit = () => {
    setTodos(
      todos.map((todo) =>
        todo.id === editId ? { ...todo, text: editText } : todo
      )
    );
    setEditId(null);
    setEditText("");
  };

  const cancelEdit = () => {
    setEditId(null);
    setEditText("");
  };

  return (
    <div className="min-h-screen bg-gray-100 py-8">
      <div className="max-w-md mx-auto bg-white rounded-lg shadow-md p-6">
        <h1 className="text-2xl font-bold text-gray-800 mb-6 text-center">
          Todo App
        </h1>
        <div className="flex mb-4">
          <input
            data-testid="inputTodo"
            type="text"
            name="textTodo"
            aria-label="textTodo"
            value={input}
            onChange={(e) => setInput(e.target.value)}
            onKeyPress={(e) => e.key === "Enter" && addTodo()}
            placeholder="Add a new todo..."
            className="flex-1 px-4 py-2 border border-gray-300 rounded-l-lg focus:outline-none focus:ring-2 focus:ring-blue-500"
          />
          <button
            onClick={addTodo}
            className="px-6 py-2 bg-blue-500 text-white rounded-r-lg hover:bg-blue-600 focus:outline-none focus:ring-2 focus:ring-blue-500"
          >
            Add
          </button>
        </div>

        <ul className="space-y-2" data-testid="ulLits">
          {todos.map((todo) => (
            <li
              key={todo.id}
              className="flex items-center gap-2 p-3 bg-gray-50 rounded-lg"
            >
              <input
                type="checkbox"
                checked={todo.completed}
                onChange={() => toggleComplete(todo.id)}
                className="w-4 h-4 text-blue-600 rounded focus:ring-blue-500"
              />
              <>
                <span
                  className={`flex-1 ${
                    todo.completed
                      ? "line-through text-gray-500"
                      : "text-gray-800"
                  }`}
                >
                  {todo.text}
                </span>
                <button
                  onClick={() => startEdit(todo.id, todo.text)}
                  className="px-3 py-1 bg-yellow-500 text-white text-sm rounded hover:bg-yellow-600"
                >
                  Edit
                </button>
                <button
                  onClick={() => deleteTodo(todo.id)}
                  className="px-3 py-1 bg-red-500 text-white text-sm rounded hover:bg-red-600"
                >
                  Delete
                </button>
              </>
            </li>
          ))}
        </ul>

        {todos.length === 0 && (
          <p className="text-gray-500 text-center mt-8">
            No todos yet. Add one above!
          </p>
        )}
      </div>
    </div>
  );
}

export default TodoApp;
