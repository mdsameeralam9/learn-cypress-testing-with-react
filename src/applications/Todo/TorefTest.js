import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import TodoApp from './TodoApp';

describe('TodoApp', () => {
  test('renders empty state', () => {
    render(<TodoApp />);
    expect(screen.getByText(/Todo App/i)).toBeInTheDocument();
    expect(screen.getByText(/No todos yet\. Add one above!/i)).toBeInTheDocument();
  });

  test('adds a todo via button', async () => {
    const user = userEvent.setup();
    render(<TodoApp />);

    const input = screen.getByPlaceholderText(/Add a new todo/i);
    await user.type(input, 'Buy milk');
    await user.click(screen.getByRole('button', { name: /Add/i }));

    expect(screen.getByText('Buy milk')).toBeInTheDocument();
  });

  test('adds a todo via Enter key', async () => {
    const user = userEvent.setup();
    render(<TodoApp />);

    const input = screen.getByPlaceholderText(/Add a new todo/i);
    await user.type(input, 'Read book{enter}');

    expect(screen.getByText('Read book')).toBeInTheDocument();
  });

  test('toggles completion', async () => {
    const user = userEvent.setup();
    render(<TodoApp />);

    const input = screen.getByPlaceholderText(/Add a new todo/i);
    await user.type(input, 'Item 1');
    await user.click(screen.getByRole('button', { name: /Add/i }));

    const checkbox = screen.getByRole('checkbox');
    expect(checkbox).not.toBeChecked();
    await user.click(checkbox);
    expect(checkbox).toBeChecked();
    expect(screen.getByText('Item 1')).toHaveClass('line-through');
  });

  test('delete removes a todo', async () => {
    const user = userEvent.setup();
    render(<TodoApp />);

    const input = screen.getByPlaceholderText(/Add a new todo/i);
    await user.type(input, 'Item 2');
    await user.click(screen.getByRole('button', { name: /Add/i }));
    await user.click(screen.getByRole('button', { name: /Delete/i }));

    expect(screen.queryByText('Item 2')).not.toBeInTheDocument();
  });

  test('edit pre-fills input and saves on Add', async () => {
    const user = userEvent.setup();
    render(<TodoApp />);

    const input = screen.getByPlaceholderText(/Add a new todo/i);
    await user.type(input, 'Old text');
    await user.click(screen.getByRole('button', { name: /Add/i }));

    await user.click(screen.getByRole('button', { name: /Edit/i }));
    // Input should be pre-filled with existing text
    expect(screen.getByDisplayValue('Old text')).toBeInTheDocument();

    // Your addTodo updates text when editId is set
    await user.clear(input);
    await user.type(input, 'New text');
    await user.click(screen.getByRole('button', { name: /Add/i }));

    expect(screen.getByText('New text')).toBeInTheDocument();
    expect(screen.queryByText('Old text')).not.toBeInTheDocument();
  });

  test('ignores empty/whitespace-only input', async () => {
    const user = userEvent.setup();
    render(<TodoApp />);

    const input = screen.getByPlaceholderText(/Add a new todo/i);
    await user.type(input, '   ');
    await user.click(screen.getByRole('button', { name: /Add/i }));

    expect(screen.getByText(/No todos yet\. Add one above!/i)).toBeInTheDocument();
  });
});
