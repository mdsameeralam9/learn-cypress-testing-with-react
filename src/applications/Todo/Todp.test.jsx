import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Todo from "./Todo";

describe("todo test", () => {
  // initial data;
  it("initial Data at UI", () => {
    render(<Todo />);
    const h1Text = screen.getByText("Todo App");
    const nodataText = screen.getByText("No todos yet. Add one above!");
    const btn = screen.getByRole("button", { name: "Add" });
    const inpt = screen.getByPlaceholderText(/Add a new todo.../);

    expect(h1Text).toBeInTheDocument();
    expect(nodataText).toBeInTheDocument();
    expect(btn).toBeInTheDocument();
    expect(inpt).toBeInTheDocument();
  });

  // add a todo by Btn click
  it("add a todo", async () => {
    const user = userEvent.setup();
    render(<Todo />);

    const int = screen.getByPlaceholderText(/Add a new todo.../);
    const btn = screen.getByRole("button", { name: "Add" });

    await user.type(int, "Buy milk");
    await user.click(btn);

    const todoText = screen.getByText("Buy milk");
    expect(todoText).toBeInTheDocument();

    const nodataText = screen.queryByText("No todos yet. Add one above!");
    expect(nodataText).not.toBeInTheDocument();
  });

  // add todo By Enter
  it("add tod by enter", async () => {
    const userevent = userEvent.setup();
    render(<Todo />);
    const inpt = screen.getByPlaceholderText(/Add a new todo.../);
    //await userevent.type(inpt, "Buy Curd{enter}");
    await userevent.type(inpt, "Buy Curd");
    await userevent.keyboard("{Enter}");
    const todoText = screen.getByText("Buy Curd");
    expect(todoText).toBeInTheDocument();
  });

  // delete todo
  it("delete item", async () => {
    const userEvnt = userEvent.setup();
    render(<Todo />);
    const inpt = screen.getByPlaceholderText("Add a new todo...");
    await userEvnt.type(inpt, "milk");
    await userEvnt.keyboard("{Enter}");

    const itemToDelete = screen.getByText("milk");
    expect(itemToDelete).toBeInTheDocument();

    const deletBtn = screen.getByRole("button", { name: "Delete" });
    await userEvnt.click(deletBtn);
  });

  // edit todo
  test("edit pre-fills input and saves on Add", async () => {
    const user = userEvent.setup();
    render(<Todo />);

    const input = screen.getByRole("textbox", { name: "textTodo" });
    await user.type(input, "Old text");
    await user.click(screen.getByRole("button", { name: /Add/i }));

    await user.click(screen.getByRole("button", { name: /Edit/i }));
    // Input should be pre-filled with existing text
    expect(screen.getByDisplayValue("Old text")).toBeInTheDocument();

    // Your addTodo updates text when editId is set
    await user.clear(input);
    await user.type(input, "New text");
    await user.click(screen.getByRole("button", { name: /Add/i }));

    expect(screen.getByText("New text")).toBeInTheDocument();
    expect(screen.queryByText("Old text")).not.toBeInTheDocument();
  });


  // testinh ul
  test("testing ul and li", async() => {
    const userEvnt = userEvent.setup();
    render(<Todo />);

    const ul = screen.getByTestId('ulLits')
    expect(ul).toBeInTheDocument()
    expect(screen.queryAllByRole('listimmmjntem')).toHaveLength(0);

    const inpt = screen.getByPlaceholderText('Add a new todo...');
    await userEvnt.type(inpt, 'buy product');
    await userEvnt.keyboard('{Enter}');

    expect(screen.getAllByRole("listitem")).toHaveLength(1)

  })

  // snapshot testing
  it("snapshot testing", () => {
    const { container } = render(<Todo />);
    expect(container).toMatchSnapshot();
  });



});
