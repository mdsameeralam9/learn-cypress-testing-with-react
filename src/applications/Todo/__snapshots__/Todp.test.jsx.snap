// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`todo test > snapshot testing 1`] = `
<div>
  <div
    class="min-h-screen bg-gray-100 py-8"
  >
    <div
      class="max-w-md mx-auto bg-white rounded-lg shadow-md p-6"
    >
      <h1
        class="text-2xl font-bold text-gray-800 mb-6 text-center"
      >
        Todo App
      </h1>
      <div
        class="flex mb-4"
      >
        <input
          aria-label="textTodo"
          class="flex-1 px-4 py-2 border border-gray-300 rounded-l-lg focus:outline-none focus:ring-2 focus:ring-blue-500"
          data-testid="inputTodo"
          name="textTodo"
          placeholder="Add a new todo..."
          type="text"
          value=""
        />
        <button
          class="px-6 py-2 bg-blue-500 text-white rounded-r-lg hover:bg-blue-600 focus:outline-none focus:ring-2 focus:ring-blue-500"
        >
          Add
        </button>
      </div>
      <ul
        class="space-y-2"
        data-testid="ulLits"
      />
      <p
        class="text-gray-500 text-center mt-8"
      >
        No todos yet. Add one above!
      </p>
    </div>
  </div>
</div>
`;
